### PR TITLE
Update our compatibility to zlib 1.2.13 and merge in remaining bug fixes.

### PR DESCRIPTION
**Pull request by mtl1979**
Requires: #1384

### CODE REVIEW

File: trees.c [hunk 1]
```diff
@@ -670,7 +670,7 @@ void Z_INTERNAL zng_tr_flush_block(deflate_state *s, char *buf, uint32_t stored_
                 opt_lenb, s->opt_len, static_lenb, s->static_len, stored_len,
                 s->sym_next / 3));
 
-        if (static_lenb <= opt_lenb)
+        if (static_lenb <= opt_lenb || s->strategy == Z_FIXED)
```
Comment (nmoinvaz) [inline trees.c:673]: The commits from madler/zlib where these changes are made should be copied over, instead of a single lump-sum commit.

Comment (mtl1979) [reply trees.c:673]: I was going to do that after GitHub CI starts working again... It had quite long backlog... I just wanted to do one CI run that actually verifies that there was no ABI breakage between 1.2.12 and 1.2.13...

Comment (mtl1979) [reply trees.c:673]: Now there is still one commit that glues things together, but I cherry-picked and cleaned the original commits for bug fixes from madler/zlib...

**Comment by nmoinvaz** [comment]
It looks like we may be missing 2 commits from Dec from here: https://github.com/madler/zlib/compare/master...develop

**Comment by mtl1979** [comment]
@nmoinvaz Are you saying we should update our compatibility to 1.2.13.1 or were those part of 1.2.13 already?

**Comment by nmoinvaz** [comment]
Actually you were right. I wasn't aware that there was an official 1.2.13 release yet. So those changes from Dec (after 1.2.13) should probably not go in until later.

**Comment by mtl1979** [comment]
> Actually you were right. I wasn't aware that there was an official 1.2.13 release yet. So those changes from Dec (after 1.2.13) should probably not go in until later.

For bug fixes, it really doesn't make a difference, but for something that breaks the ABI or API, it's completely different story... Some of our functions, like deflateBound(), have diverged so much that cherry-picking is quite pointless as whole commit needs to be rewritten.

File: deflate.c [hunk 2]
```diff
@@ -660,6 +660,12 @@ unsigned long Z_EXPORT PREFIX(deflateBound)(PREFIX3(stream) *strm, unsigned long
         wraplen = ZLIB_WRAPLEN;
     }
 
+    if (s->level == 0) {
```
Comment (nmoinvaz) [inline deflate.c]: This if statement could go inside the if statement on line 670 if we wanted.

**Comment by mtl1979** [comment]
Rebased.

**Review by nmoinvaz — APPROVED** [review]
LGTM. Probably want to rebase for abi package changes.

**Comment by mtl1979** [comment]
> LGTM. Probably want to rebase for abi package changes.

More likely if #1412 is merged before this as this was already rebased after ABI fixes, even though this didn't need recreating ABI files as different base commit is used.

**Comment by Dead2** [comment]
Rebase?

**Comment by mtl1979** [comment]
> Rebase?

Done... Just to see everything still passes with flying colours...

**Review by Dead2 — APPROVED** [review]

**Pull request merged**